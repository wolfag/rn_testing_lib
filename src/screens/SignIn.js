import { useNavigation } from '@react-navigation/native';
import React, { useState } from 'react';
import { KeyboardAvoidingView, StyleSheet, Text } from 'react-native';
import { Button, ErrorText, Input } from '../components/Form';

const useLoginFormState = () => {
  const navigation = useNavigation();
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const [submit, setSubmit] = useState(false);

  let isUsernameValid = false;
  let isPasswordValid = false;
  let usernameErrorMsg;
  let passwordErrorMsg;

  if (username === 'example') {
    isUsernameValid = true;
  }

  if (password === 'asdf') {
    isPasswordValid = true;
  }

  if (submit && !isUsernameValid) {
    usernameErrorMsg = 'Invalid username.';
  }

  if (submit && !isPasswordValid) {
    passwordErrorMsg = 'Invalid password.';
  }

  const handleSubmit = () => {
    setSubmit(true);

    if (!isUsernameValid || !isPasswordValid) return;

    fetch('https://jsonplaceholder.typicode.com/users', {
      method: 'POST',
      body: JSON.stringify({ username, password }),
    })
      .then((res) => res.json())
      .then(() => {
        navigation.push('App');
      })
      .catch((error) => {
        console.log({ error });
      });
  };

  return {
    username: {
      value: username,
      set: setUsername,
      valid: isUsernameValid,
      error: usernameErrorMsg,
    },
    password: {
      value: password,
      set: setPassword,
      valid: isPasswordValid,
      error: passwordErrorMsg,
    },
    submit: {
      value: submit,
      go: handleSubmit,
    },
  };
};

export default function SignIn() {
  const { username, password, submit } = useLoginFormState();

  return (
    <KeyboardAvoidingView style={styles.container} behavior='position'>
      <Text style={styles.headerText}>Login</Text>
      <Input
        label='Username'
        placeholder='example'
        onChangeText={username.set}
        error={username.error}
        testID='SignIn.usernameInput'
      />
      <Input
        label='Password'
        placeholder='***'
        secureTextEntry
        onChangeText={password.set}
        error={password.error}
        testID='SignIn.passwordInput'
      />
      <ErrorText messages={[username.error, password.error]} />
      <Button testID='SignIn.Button' text='Login' onPress={submit.go} />
    </KeyboardAvoidingView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    justifyContent: 'center',
    paddingHorizontal: 40,
  },
  headerText: {
    color: '#353031',
    fontWeight: 'bold',
    fontSize: 34,
    marginBottom: 10,
  },
});
