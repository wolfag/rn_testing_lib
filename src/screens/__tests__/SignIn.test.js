import React from 'react';
import { render, fireEvent, act } from '@testing-library/react-native';

import SignIn from '../SignIn';

const flushMicrotasksQueue = () => new Promise((res) => setImmediate(res));

const mockNavigation = jest.fn();
jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    push: mockNavigation,
  }),
}));

describe('SignIn screen', () => {
  beforeEach(() => {
    mockNavigation.mockClear();
  });

  it('render default elements', () => {
    const { getAllByText, getByPlaceholderText } = render(<SignIn />);

    expect(getAllByText('Login').length).toBe(2);
    getByPlaceholderText('example');
    getByPlaceholderText('***');
  });

  it('show invalid inputs messages', () => {
    const { getByTestId, getByText } = render(<SignIn />);

    fireEvent.press(getByTestId('SignIn.Button'));
    getByText('Invalid username.');
    getByText('Invalid password.');
  });
  it('show invalid user name error messages', () => {
    const { getByTestId, getByText, queryAllByText } = render(<SignIn />);

    fireEvent.changeText(getByTestId('SignIn.passwordInput'), 'asdf');
    fireEvent.press(getByTestId('SignIn.Button'));
    getByText('Invalid username.');
    expect(queryAllByText('Invalid password.').length).toBe(0);
  });

  it('show invalid password error messages', () => {
    const { getByTestId, getByText, queryAllByText, debug } = render(
      <SignIn />,
    );

    fireEvent.changeText(getByTestId('SignIn.usernameInput'), 'example');
    fireEvent.press(getByTestId('SignIn.Button'));
    getByText('Invalid password.');
    expect(queryAllByText('Invalid username.').length).toBe(0);

    debug();
  });

  it('handle valid input submission', async () => {
    fetch.mockResponseOnce(JSON.stringify({ passes: true }));

    const { getByTestId } = render(<SignIn />);

    fireEvent.changeText(getByTestId('SignIn.usernameInput'), 'example');
    fireEvent.changeText(getByTestId('SignIn.passwordInput'), 'asdf');

    fireEvent.press(getByTestId('SignIn.Button'));

    expect(fetch.mock.calls).toMatchSnapshot();
    await act(flushMicrotasksQueue);

    expect(mockNavigation).toBeCalledWith('App');
  });
});
