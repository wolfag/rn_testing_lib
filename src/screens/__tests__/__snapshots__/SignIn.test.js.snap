// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn screen handle valid input submission 1`] = `
Array [
  Array [
    "https://jsonplaceholder.typicode.com/users",
    Object {
      "body": "{\\"username\\":\\"example\\",\\"password\\":\\"asdf\\"}",
      "method": "POST",
    },
  ],
]
`;
